Add Enterprise CommCare Client Version Compliance Report

jingcheng16 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

New Tile
<img width="1438" alt="image" src="https://github.com/user-attachments/assets/17bd41e1-2b6a-48e0-921a-e700993534a1">

Report
<img width="678" alt="image" src="https://github.com/user-attachments/assets/cad5eaf0-0829-4c13-8df8-db22366a6bc9">

API response
<img width="1436" alt="image" src="https://github.com/user-attachments/assets/cf81bcd3-9408-47a6-8462-754c0fb7c095">


## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-16236

I'm a bit worried about the performance. When I test on production, it can take more than 2 minutes to generate total number for account like CRS

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
Doing iterative release, so no FF. Once it is merged, we release it.

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally, on staging, and also on production.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Add test for util function I wrote

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
QA Ticket: https://dimagi.atlassian.net/browse/QA-7315

Verify the tile works as expected:

- An enterprise account without any mobile workers should see '--' shown up in tile.

Email the report:
- A mobile worker doesn't have any device yet, and haven't submitted any forms yet, will not be included in the report
- A mobile worker whose latest submission is using the latest commcare version, will **not be included** in the report. 

- A mobile worker whose latest submission is using an older commcare version, ([instruction to install an older version](https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/2143946363/Offline+Install+of+CommCare+on+Android)), this mobile worker **will be included** in the report. Then use the same mobile worker, but submitted a form using the "Log in as xx mobile worker" feature, verify this mobile worker **is not** in the report. Because I assume in this case we're always using the latest commcare version. (Using the same mobile worker is to verify the submission through "Log in as xx mobile worker" feature will replace the previous submission when calculate version number)

- Can a mobile worker submitted form using script? If yes, please verify if the latest submission is by script, this mobile worker **is not** in the report. Because I assume in this case we're always using the latest commcare version.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
Refactor: rename function

jingcheng16 committed:
Util to tell if a version is out of date

jingcheng16 committed:
Add Commcare Version Compliance Tile

jingcheng16 committed:
Add CommCare Version Compliance API

jingcheng16 committed:
Refactor: move get_report_task to the base class

jingcheng16 committed:
Show -- for the total if the account has no mobile worker yet

mjriley commented on corehq/apps/builds/utils.py:
```diff
@@ -40,3 +40,47 @@ def extract_build_info_from_filename(content_disposition):
+    Get the latest CommCare version that was available at a given time.
+    Excludes superuser-only versions.
+    Menu items are already in chronological order (newest last).
+    If no target time is provided, return the latest version at now.
```
change to: return the latest version available now

jingcheng16 replied:
a65c782bd982626d2c8526be4125bde6a5b0d26b

mjriley commented on corehq/apps/builds/utils.py:
```diff
@@ -40,3 +40,47 @@ def extract_build_info_from_filename(content_disposition):
+    return None
+
+
+def _parse_version(version_str):
+    """Convert version string to comparable tuple"""
+    if version_str:
+        try:
+            return tuple(int(n) for n in version_str.split('.'))
+        except (ValueError, AttributeError):
+            return None
+    return None
```
Nit: move beneath the function that uses it

jingcheng16 replied:
a44b2002dac682103c2027238e0a5111ca72427d

mjriley commented on corehq/apps/builds/tests/test_utils.py:
```diff
@@ -0,0 +1,48 @@
+
+
+class TestVersionUtils(SimpleTestCase):
+    def test_parse_version(self):
+        test_cases = [
+            ('2.53.0', (2, 53, 0)),
+            ('2.53', (2, 53)),
+            ('1.0.0', (1, 0, 0)),
+            ('7', (7,)),
+            ('', None),
+            (None, None),
+            ('2.53.0.invalid', None),
+            ('not_a_version', None),
+        ]
+
+        for version, expected in test_cases:
+            with self.subTest(version=version, expected=expected):
+                result = _parse_version(version)
+                self.assertEqual(
+                    result,
+                    expected,
+                    f"Failed to parse '{version}'. Expected {expected}, got {result}"
+                )
```
Test can be removed

jingcheng16 replied:
3016d9e0505c907532e77a036bec54e81afeb813

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -7,9 +7,11 @@
 
 from memoized import memoized
 
+from corehq.apps.reports.standard.deployments import _format_commcare_version
```
It's a smell to import a private function. Should this now be a public function?

jingcheng16 replied:
c25a980ec516215fb65b406299624c21a66dbd46

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -383,3 +388,68 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+
+
+class EnterpriseCommCareVersionReport(EnterpriseReport):
+    title = _('CommCare Client Version Compliance')
+    metric = _('%% of mobile workers on the latest commcare client version ')
```
change to lazy translation

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -383,3 +388,68 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+
+
+class EnterpriseCommCareVersionReport(EnterpriseReport):
+    title = _('CommCare Client Version Compliance')
+    metric = _('%% of mobile workers on the latest commcare client version ')
```
Metric may be confusing as it may suggest Kissmetrics or Google Analytics. We may want to rename it? Description, subtitle, etc.

jingcheng16 replied:
5f91e2b1109de4e49d6cd9cb41b39fff2d60ed0e

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -383,3 +388,68 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+            _('Project Space'),
+            _('Latest Version Available at Submission'),
+            _('Version in Use'),
+
```
Nit: remove space

jingcheng16 replied:
a65c782bd982626d2c8526be4125bde6a5b0d26b

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -383,3 +388,68 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+    def rows_for_domain(self, domain_obj):
+        rows = []
+
+        for user in get_all_user_rows(domain_obj.name,
+                                    include_web_users=False,
+                                    include_mobile_users=True,
+                                    include_inactive=False,
+                                    include_docs=True):
```
Nit: tab formatting

jingcheng16 replied:
a65c782bd982626d2c8526be4125bde6a5b0d26b

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -383,3 +388,68 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+                                    include_inactive=False,
+                                    include_docs=True):
+            user = CouchUser.wrap_correctly(user['doc'])
+            last_submission = user.reporting_metadata.last_submission_for_user
```
Possibly could be turned into its own function, like: `get_last_version_for_user(user)`. This could be especially important if this same logic is used in the Application Status Report.

jingcheng16 replied:
2e874e57c19bd581866277a544408ff90c7bc6a7

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -383,3 +388,68 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+                total_mobile_workers += domain_mobile_workers
+                total_up_to_date += domain_mobile_workers - len(self.rows_for_domain(domain_obj))
+        percentage = (total_up_to_date / total_mobile_workers) if total_mobile_workers else 0
+        return f"{percentage * 100:.0f}%"
```
Change to 1 decimal place

jingcheng16 replied:
c9586dc2534329070e385a95e7bd1914a8b379a7
I also make it a function to format the percentage, so both you and I can reuse it, thus making sure the percentage will always have just one digit.

mjriley replied:
Nice, thanks!

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -383,3 +388,68 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+            domain_mobile_workers = get_mobile_user_count(domain_obj.name, include_inactive=False)
+            if domain_mobile_workers:
+                total_mobile_workers += domain_mobile_workers
+                total_up_to_date += domain_mobile_workers - len(self.rows_for_domain(domain_obj))
```
There is definitely a performance concern here, as if this summary query takes 2 minutes, that is 2 minutes blocking a web worker. It's worth trying to see if Elasticsearch is any faster here. Otherwise, we may want to cache this statistic.

jingcheng16 replied:
1700c2f6d664f5435b938324e2ca24365ac560fc

jingcheng16 replied:
Tried on production again, it takes between 2 minutes and 3 minutes. Last time I was about to leave my apartment so I didn't wait until it finished, so I guess last time when I query couch directly maybe it longer than 3 minutes. There is improvement by using elastic search. Is the improvement acceptable? I think maybe yes? Because not every enterprise account would have that amount of mobile workers...

mjriley replied:
If we're talking minute I think we need to think about caching or turning this into a celery task. I can imagine that multiple requests to this page could cause a disruption in our responsiveness that would be hard to diagnose

mjriley commented on corehq/apps/enterprise/static/enterprise/js/project_dashboard.js:
```diff
@@ -175,7 +175,11 @@ hqDefine("enterprise/js/project_dashboard", [
             url: initialPageData.reverse("enterprise_dashboard_total", slug),
             success: function (data) {
-                $display.html(Number(data.total).toLocaleString());
+                if (typeof data.total === 'number') {
+                    $display.html(Number(data.total).toLocaleString());
+                } else {
+                    $display.html(data.total);
+                }
```
If this is intended to support percentages, do we need to be concerned about locale formatting (i.e. using ',' for a decimale separator) for those as well?

jingcheng16 replied:
Cherry picked your commit in PR #35480 in 1b67a86b55f8e73d3145e7e64c42be7a8c2f7d49

mjriley reviewed (changes requested):
Most of the comments are non-blocking, but the most serious was just the performance concern for the summary tile. If Elasticsearch turns out to not be an option, or if that option is still too slow, we'd like to change this so that the data is cached, with the theory that we can stomach blocking our webworkers for 2 minutes every once in a while, but don't want to do so repeatedly.

mjriley commented:
Instead of caching (which still could generate some negative effects if multiple enterprises all hit the enterprise console at the same time), we could also look at changing the system entirely. Right now, the enterprise console sends out a request for each tile and essentially blocks until the webworker processes it. Instead, we could send an initial request to the webworker which offloaded the task to celery, and the page would then poll the webworker to see if the celery task had completed. Not something we need to do here, but its an option in general for this and other tiles

jingcheng16 committed:
Move is_out_of_date above _parse_version

jingcheng16 committed:
change text and fix format

jingcheng16 committed:
Remove unnecessary code

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -383,3 +388,68 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+    metric = _('%% of mobile workers on the latest commcare client version ')
+
+    def __init__(self, account, couch_user, **kwargs):
+        super().__init__(account, couch_user, **kwargs)
```
3016d9e0505c907532e77a036bec54e81afeb813

jingcheng16 committed:
Change private function to public since it's imported by other file

jingcheng16 committed:
delay the translation until rendering

jingcheng16 committed:
Rename metric to description

jingcheng16 committed:
Percentage in enterprise console will have one digit

jingcheng16 committed:
Use elastic search

jingcheng16 committed:
Having a util function to handle the commcare version the mobile worker is in use

jingcheng16 committed:
isort

jingcheng16 committed:
Remove formatting of version in enterprise console so it's comparable

jingcheng16 committed:
ES returning dates in iso format so add conversion to avoid type error

jingcheng16 commented:
@mjriley @biyeun Addressed all feedback and this is ready for review again.

mjriley committed:
Merge branch 'master' into jc/commcare_client_version_compliance_report

jingcheng16 committed:
flake8

jingcheng16 commented:
@mjriley Thank you for helping me resolving the merge issue. I have updated all my reply with the new SHA and it is ready for review.

jingcheng16 committed:
Change title

jingcheng16 committed:
Merge branch 'master' into jc/commcare_client_version_compliance_report

jingcheng16 committed:
Rename description to total_description

jingcheng16 committed:
Update label....

mjriley commented on corehq/apps/reports/standard/deployments.py:
```diff
@@ -352,11 +355,10 @@ def process_rows(self, users, fmt_for_export=False):
-            else:
-                if device and device.get('commcare_version', None):
-                    commcare_version = _get_commcare_version(device['commcare_version'])
+            commcare_version, last_seen = get_commcare_version_and_date_from_last_usage(last_sub,
```
It looks like the prior version of this function only set `last_seen` once, and that line still exists below (where its being set to `last_sub[submission_date]`. Are we setting `last_seen` here only to overwrite it? Should we ignore the `last_seen` return value from this function and rely on the other line? Or can we generalize this logic so that we remove the old line and replace it with the result of this function?

jingcheng16 replied:
Thanks for catching it. This is my oversight. 0d85d21970c4ab93bfffe1300cee233978fb8027
The logic cannot be generalized, because here the `last_seen` will only be last submission time, while the date returned by `get_commcare_version_and_date_from_last_usage` can be fall back to last_device.

mjriley commented on corehq/apps/reports/util.py:
```diff
@@ -851,3 +851,29 @@ def domain_copied_cases_by_owner(domain, owner_ids):
+    date_of_use = None
+
+    if last_submission and last_submission.get('commcare_version'):
+        version = last_submission.get('commcare_version')
+        version_in_use = format_commcare_version(version) if formatted else version
```
I know most of this was copied from the previous implementation, but there is a bit of duplication for handling the version. Is the following pseudocode better?
```
if ...
    version = last_submission.get('commcare_version')
    ...
elif ...
   version = last_device.get('commcare_version')
   ...

if version:
  version_in_use = format_commcare_version(version) if formatted else version
```

an improvement?

jingcheng16 replied:
Thank you! 34e358b6c09fb54a18f09b7b348a4cb520870498

mjriley reviewed (commented):
Thanks for the changes. I think the initial review is still largely the feeling -- I think the code itself is solid, but I'm concerned about a long-running task blocking our webworkers

biyeun committed:
adjust tile layout

jingcheng16 committed:
Merge branch 'master' into jc/commcare_client_version_compliance_report

jingcheng16 committed:
Rename ignored variable

The underscore is used as an alias for gettext translation function, so we can't use it to ignore the second return value

jingcheng16 committed:
Refactor: format version only once

jingcheng16 committed:
Memoize get_latest_version_at_time and use minute precision for date_of_use

jingcheng16 committed:
Fix UnboundLocalError

local variable 'version' referenced before assignment

jingcheng16 committed:
Remove microseconds too

jingcheng16 committed:
Fetch CommCareBuildConfig only once

This is the main breakthrough of the performance issue. When inspecting the profiling results, we noticed the bottleneck is _pickle.loads. CommCareBuildConfig is caching a whole couch document, which takes up a long time for pickle to deserialization.

jingcheng16 committed:
memoize build time to avoid repetitively fetching build for each version

biyeun committed:
update naming convention for elements refrenced in javascript
best practice is not to reference style classes as logic elements in javascript. additionally css classes referenced by javascript should be prefixed with js-

jingcheng16 committed:
This tile don't need domain_obj

It saves memory also saves 1 second for large enterprise account

jingcheng16 committed:
Move total above rows_for_domain

jingcheng16 committed:
Use gevent pooling for parallel domain processing

The latest performance bottlenecks primarily in I/O operations (specifically
in BufferedReader.readline).

mjriley commented on corehq/apps/builds/utils.py:
```diff
@@ -40,3 +44,61 @@ def extract_build_info_from_filename(content_disposition):
+    return None
+
+
+@memoized
```
The use of `memoized` makes me nervous, as my understanding is that the cached value will stick around until it is explicitly cleared, or the machine restarts. Say we created a version, this ran for that version, and then we had to delete the version. This function would still return the old build time, despite the version no longer existing, right? Worse still, because the cache is local to the machine, one webworker might return the old value, while a different webworker, that hadn't previously processed the request, would return the current value. Does it seem safer to perform manual, per-request caching? I.e. create a dictionary and insert values as the request generates them, but throw that dictionary away at the end of the request.

jingcheng16 replied:
Thank you for noticing it! [I verified in our doc](https://commcare-hq.readthedocs.io/caching_and_memoization.html#lifecycle):
> Memoized on any other function/callable:
The cache lives on the callable, so if it’s globally scoped and never gets garbage collected, neither does the cache


So I'm changing it to a quickcache, and clear it every 10 minutes: 44d9db5

mjriley replied:
Is the intention 10 minutes? `100 * 60` seems to be 100 minutes, right? I'd still prefer a completely local, request-based caching mechanism, because I think it's more in-line with what you're trying to do, but I'll defer to your judgment here. This code works. The downside is that A) caching is harder to debug, B) quickcache is making redis/network requests, and C) we can still run into that inconsistent state due to a stale cache, but now that stale cache will expire after the timeout.

jingcheng16 replied:
Hi Matt, my initial intention is 10 minutes. But now I think 100 minutes is better. I think we will only run into inconsistent state if we release a version then delete it, which I believe this rarely happen. So no need to worry about (A) and (C). By setting `memoize_timeout`, we reduce Redis requests by serving cached values directly from memory if the same process makes repeated calls.

jingcheng16 replied:
I changed my mind!!! Quickcache added a lot of overhead and resulted in worse performance issue, so I implemented a dictionary to store the cache as you suggested: 83aecc62dbf1d35ee75654fc405b673ad7bc1024

mjriley replied:
Love it! Daniel did mention to me, however, that the problem with the initial `@memoized` solution wasn't the caching so much as the _global_ caching. Rather than passing the cache in explicitly, and having to write the logic to insert `None` into the cache when an exception is thrown, for example, it might be cleaner to create an object that can limit the scope of the memoization. But entirely optional -- what you implemented is what I had asked for

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -405,6 +422,94 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+
+        return _format_percentage_for_enterprise_tile(total_up_to_date, total_mobile_workers)
+
+    def rows_for_domain(self, domain_name, config):
```
Nit: I believe we refer to a domain object as `domain_obj` and the domain name as `domain`

jingcheng16 replied:
7599675a56c9820eb540ff0802cf236a9e549f4e

mjriley commented on corehq/apps/reports/util.py:
```diff
@@ -851,3 +851,31 @@ def domain_copied_cases_by_owner(domain, owner_ids):
+    if version:
+        version_in_use = format_commcare_version(version) if formatted else version
+
+    return version_in_use, date_of_use
```
Would it be possible to convert `date_of_use` from a string to a datetime here? It doesn't look like `date_of_use` is referenced by the existing report functions, and this would simplify the logic in `enterprise`, as we could remove the logic that checks whether this is a date or a string

jingcheng16 replied:
Thank you. I agree with you!! 27aa65f30769b562b0d5f19f1b51fa8f50045ffa

mjriley commented on corehq/apps/enterprise/static/enterprise/js/project_dashboard.js:
```diff
@@ -270,13 +270,13 @@ hqDefine("enterprise/js/project_dashboard", [
 
         kissmetrics.track.event(`[${metricType}] Visited page`);
-        $(".report-panel").each(function () {
+        $(".js-report-panel").each(function () {
```
@biyeun I did a brief search, but can you recommend any reading on this practice?

mjriley reviewed (commented):
Great findings and improvements. Before I approve this, I just want to make sure our performance optimizations are worth it and don't have negative side effects. I left commends regarding our use of `memoized` and running the operations in parallel.

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -405,6 +422,94 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+
+            # Remove seconds and microseconds to reduce the number of unique timestamps
+            # This helps with performance because get_latest_version_at_time is memoized
+            if isinstance(date_of_use, str):
+                date_of_use = datetime.strptime(date_of_use, ISO_DATETIME_FORMAT)
+            date_of_use_minute_precision = date_of_use.replace(second=0, microsecond=0)
+
+            latest_version_at_time_of_use = get_latest_version_at_time(config, date_of_use_minute_precision)
```
Seconds and microseconds are removed from the form submission time, not our build. So the result is, for example, if we release a new version `Version 3.3` at 13:10:30, a mobile worker submit a form using `Version 3.2` in 13:10:40, after removing seconds, it is become 13:10:00, the latest version at 13:10:00 is still `Version 3.2`, so this mobile worker won't be considered as using the out-of-date commcare version, although technically he is using out-of-date commcare version. I'm sacrificing a bit accuracy for caching the result of `get_latest_version_at_time`.

mjriley replied:
If we choose to go forward with this approach, I'd want it documented that we're reducing the accuracy of results in order to improve performance. I'm not sure reducing the accuracy of report data for the sake of a quicker summary makes sense. However, can we avoid this entirely now?  It looks like the main bottlenecks for this function were fetching the config and build time, but both of those things have now been moved into their own areas. The config is only fetched once, outside of `get_latest_version_at_time`, and `get_build_time` is now cached. So I don't think we need to do any caching on `get_latest_version_at_time` anymore. At least on staging, our config has 65 items. I don't think it should be expensive to iterate backward through those items until we find a build time less than our target.

jingcheng16 replied:
Right. It is removed in 83aecc62dbf1d35ee75654fc405b673ad7bc1024

mjriley replied:
Thank you!

jingcheng16 committed:
Refer domain name as domain and fix a format in template

jingcheng16 commented on corehq/apps/enterprise/templates/enterprise/project_dashboard.html:
```diff
@@ -15,7 +15,9 @@
     {% for report in reports %}
-      {% include 'enterprise/partials/project_tile.html' with report=report uses_date_range=uses_date_range %}
+      <div class="col-md-6 col-lg-6 col-xl-3 mb-3">
+        {% include 'enterprise/partials/project_tile.html' with report=report uses_date_range=uses_date_range%}
```
7599675a56c9820eb540ff0802cf236a9e549f4e

jingcheng16 committed:
quickcache on globally scoped function

so it can be cleared when timeout

jingcheng16 committed:
keep exception handling as close as possible to the operation that can raise the exception

jingcheng16 commented on corehq/apps/builds/utils.py:
```diff
@@ -40,3 +44,61 @@ def extract_build_info_from_filename(content_disposition):
+    build = CommCareBuild.get_build(version, latest=True)
+    if build and build.time:
+        return build.time
+    return None
```
Thank you! Sorry this is being extracted into its own function and memoized in a rush. Fixed in bb87820

mjriley replied:
Nit: Your version is an improvement, and it hides the exception complexity, but it still does more than is needed in the try/except block. `CommCareBuild` is the statement that can throw, so its slightly better and more explicit to write:
```python
try:
  build = CommCareBuild.get_build(version, latest=True)
except KeyError:
  return None
return build.time if build and build.time else None
```

because it scopes the exception block to _just_ the code that can raise the exception. I also prefer to return an explicit value, rather than letting it implicitly return nothing.

jingcheng16 replied:
84f1080d10515d9361af2478937078c172128a93

jingcheng16 committed:
Remove usage of gevent due to es thread pool size concern

jingcheng16 committed:
Always return datetime object in util function

jingcheng16 committed:
Moved exception handling closer to the operation that may raise it

gherceg commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -464,3 +569,18 @@ def rows_for_domain(self, domain_obj):
+        pool.join()
+        results.extend(chunk_results)
+
+    return results
```
Thanks for raising this concern Matt.

> If each of them tries to max out their elasticsearch connections, it seems like we're basically back to running operations in sequence, rather than in parallel?

I'm understanding this as a hypothetical right? Basically saying if we were to use this same technique on every tile, we would end up waiting for I/O elsewhere in the chain? But there is potentially still a world where we employ this technique on tile queries where it is especially useful? We don't have to necessarily max out the connection limit for this worker as I imagine there are still benefits to chunking this by as little as 2 or 3.

And just to reiterate, the larger concern I have is not with this performance change being moot, but it actually being harmful in that it could max out elasticsearch's search pool queue, as Jing noted, resulting in errors.

mjriley replied:
Right, my concern was with adding needless complexity, if these performance improvements didn't turn out to be performance improvements. It sounds like you're concerned that they could lead to system stability issues, which is definitely more pressing. We definitely want to preserve the ability to selectively optimize specific queries, but I think we'd want to specifically call that out here with comments. Otherwise, I think there's a pretty high chance that this code gets copied as an example of how we write tile queries, and it leads to the situation I outlined, where every tile requests max capacity, resulting in no one really getting performance improvements

jingcheng16 committed:
Implement per-request cache dictionary to avoid quickcache overhead

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -405,6 +416,92 @@ def _get_individual_export_rows(self, exports, export_line_counts):
+        config = CommCareBuildConfig.fetch()
+        version_cache = {}
+
+        def total_for_domain(domain):
+            mobile_workers = get_mobile_user_count(domain, include_inactive=False)
+            if mobile_workers == 0:
+                return 0, 0
+            outdated_users = len(self.rows_for_domain(domain, config, version_cache))
+            return mobile_workers, outdated_users
```
I feel like nested functions generally make the code more confusing. One good use for inner functions is when the inner functions depends on a value from the outer function (i.e. closures). A passable reason would be if the function is extremely simple (a one-liner) and creating an inner function improves readability. I don't think this function meets either of those criteria, though. Nesting it this way means it becomes harder to debug the nested function.

mjriley replied:
Oh...I guess it technically is a closure, as it uses the `version_cache`. I guess the choice to nest this function was to avoid having `total_for_domain` take the `version_cache` parameter?

jingcheng16 replied:
My initial intention is not avoid having total_for_domain take the version_cache parameter. But I think you just find another reason to next it. :joy: I nest it in total because, otherwise total_for_domain looks like I want to override the total_for_domain in the base class. While I have no intention to override it, I will compute total in a special way in this tile.

mjriley commented on corehq/apps/reports/util.py:
```diff
@@ -851,3 +852,34 @@ def domain_copied_cases_by_owner(domain, owner_ids):
+    if version:
+        version_in_use = format_commcare_version(version) if formatted else version
+
+    if isinstance(date_of_use, str):
+        date_of_use = datetime.strptime(date_of_use, ISO_DATETIME_FORMAT)
```
Is this check necessary? I think `date_of_use` will always be a string here.  I think the conditional suggests that sometimes it won't be a string, which is confusing if that isn't true.

jingcheng16 replied:
82bb7616018d97ba7eec526d941a0595f20cdf53

mjriley commented on corehq/apps/builds/utils.py:
```diff
@@ -40,3 +43,71 @@ def extract_build_info_from_filename(content_disposition):
+    if not target_time:
+        return config.get_default().version
+
+    if isinstance(target_time, str):
+        target_time = datetime.strptime(target_time, ISO_DATETIME_FORMAT)
```
These lines are now unnecessary with the latest changes, right? We can be assured that `target_time` will be a `datetime` object?

jingcheng16 replied:
82bb7616018d97ba7eec526d941a0595f20cdf53

mjriley reviewed (approved):
Thanks for the updates! I think we're at a point where the current code is appropriate for release, so approving.

jingcheng16 committed:
Remove unnecessary code

1. ES return date in string
2. get_latest_version_at_time now will only take datetime object for target_time
